### PR TITLE
Improve parameter-binding performance of large instructions

### LINKED ISSUE #10282: Parameter binding substantially slower on wrapped circuit vs unwrapped
### What should we add?

Calling `QuantumCircuit.bind_parameters` can have drastically different performance depending on whether or not the circuit contains wrapped instructions. For example, comparing the time to build and bind parameters for an `EfficientSU2` circuit, where in the first case, we call `bind_parameters` directly on `EfficientSU2`, and in the second, after unwrapping `EfficientSU2` via a call to `decompose`:
```python
from simple_benchmark import benchmark

def _build_wrapped(num_layers):
    qc = EfficientSU2(100, entanglement='linear', reps=int(num_layers))
    qc.bind_parameters({p: math.pi/2 for p in qc.parameters})
    
def _build_unwrapped(num_layers):
    qc = EfficientSU2(100, entanglement='linear', reps=int(num_layers)).decompose()
    qc.bind_parameters({p: math.pi/2 for p in qc.parameters})
    
funcs = [_build_wrapped, _build_unwrapped]
b = benchmark([_build_wrapped, _build_unwrapped],
              {i: i for i in np.linspace(1, 100, 5)},
              'num_layers',
              function_aliases={_build_wrapped: 'wrapped', _build_unwrapped: 'decomposed'}
             )
```

image

Related to https://github.com/Qiskit/qiskit-terra/pull/10269 . During profiling @mtreinish found most of the time for the wrapped case spent in recursive calls to https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/circuit/quantumcircuit.py#L2933-L2944 .

### LINKED ISSUE #10283: Circuit parameter binding does not bind `global_phase` in instruction definitions
### Environment

- **Qiskit Terra version**: 0.24.1
- **Python version**: 3.10
- **Operating system**: macOS 13.3.1

### What is happening?

When a custom gate with a set `definition` contains a parametrised global phase, any circuit that contains that custom gate will fail to bind the global phase during parameter assignment.

### How can we reproduce the issue?

```python
import math

from qiskit.circuit import Parameter, QuantumCircuit
from qiskit.quantum_info import Operator

x = Parameter("x")

custom = QuantumCircuit(1, global_phase=x).to_gate()
base = QuantumCircuit(1)
base.append(custom, [0], [])

Operator(base.assign_parameters({x: math.pi}))
```
```
TypeError                                 Traceback (most recent call last)
Cell In[1], line 12
      9 base = QuantumCircuit(1)
     10 base.append(custom, [0], [])
---> 12 Operator(base.assign_parameters({x: math.pi}))

File ~/code/qiskit/terra/qiskit/quantum_info/operators/operator.py:85, in Operator.__init__(self, data, input_dims, output_dims)
     76     self._data = np.asarray(data, dtype=complex)
     77 elif isinstance(data, (QuantumCircuit, Operation)):
     78     # If the input is a Terra QuantumCircuit or Operation we
     79     # perform a simulation to construct the unitary operator.
   (...)
     83     # conditional gates, measure, or reset will cause an
     84     # exception to be raised.
---> 85     self._data = self._init_instruction(data).data
     86 elif hasattr(data, "to_operator"):
     87     # If the data object has a 'to_operator' attribute this is given
     88     # higher preference than the 'to_matrix' method for initializing
     89     # an Operator object.
     90     data = data.to_operator()

File ~/code/qiskit/terra/qiskit/quantum_info/operators/operator.py:614, in Operator._init_instruction(cls, instruction)
    612 if isinstance(instruction, QuantumCircuit):
    613     instruction = instruction.to_instruction()
--> 614 op._append_instruction(instruction)
    615 return op

File ~/code/qiskit/terra/qiskit/quantum_info/operators/operator.py:691, in Operator._append_instruction(self, obj, qargs)
    689 else:
    690     new_qargs = [qargs[bit_indices[tup]] for tup in instruction.qubits]
--> 691 self._append_instruction(instruction.operation, qargs=new_qargs)

File ~/code/qiskit/terra/qiskit/quantum_info/operators/operator.py:669, in Operator._append_instruction(self, obj, qargs)
    666 if obj.definition.global_phase:
    667     dimension = 2**obj.num_qubits
    668     op = self.compose(
--> 669         ScalarOp(dimension, np.exp(1j * float(obj.definition.global_phase))),
    670         qargs=qargs,
    671     )
    672     self._data = op.data
    673 flat_instr = obj.definition

File ~/code/qiskit/terra/qiskit/circuit/parameterexpression.py:480, in ParameterExpression.__float__(self)
    478 except (TypeError, RuntimeError) as exc:
    479     if self.parameters:
--> 480         raise TypeError(
    481             "ParameterExpression with unbound parameters ({}) "
    482             "cannot be cast to a float.".format(self.parameters)
    483         ) from None
    484     try:
    485         # In symengine, if an expression was complex at any time, its type is likely to have
    486         # stayed "complex" even when the imaginary part symbolically (i.e. exactly)
    487         # cancelled out.  Sympy tends to more aggressively recognise these as symbolically
    488         # real.  This second attempt at a cast is a way of unifying the behaviour to the
    489         # more expected form for our users.
    490         cval = complex(self)

TypeError: ParameterExpression with unbound parameters ({Parameter(x)}) cannot be cast to a float.
```

### What should happen?

All parameter usage in custom definitions should be bound.

### Any suggestions?

This is due to ad-hoc recursion/iteration in `QuantumCircuit._rebind_definition`.  It treats `Instruction._definition` like a `list[tuple[Instruction, qubits, clbits]]`, which (a long time ago) was a possibility.  Since `Instruction._definition` is now required to be a `QuantumCircuit`, the simplest fix is to make the definition-rebinding an explicit recursion via `QuantumCircuit.assign_parameters`.

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

Previously, the parameter-assignment methods of `QuantumCircuit` had poor performance when an instruction had a complex definition that involved many of the parameters being bound.  The strategy of binding each parameter separately led to each definition being copied and rebound multiple times, with each rebinding being recursive all the way down.

This commit makes the definition rebinding happen only once per instruction, and updates the data model used to make it a complete recursion through `QuantumCircuit.assign_parameters`.  This has the side effect of fixing an issue where internal global phases would not be updated.

The algorithmic change that enables this (just rebind the definition at the end) is rather simpler than the length of this patch suggests.  This is just because the previous structure of separating out a single `_assign_parameter` method made it harder to restructure the logic without introducing unpleasant stateful coupling between the driver and helper methods.  Instead, I inlined most of the helper functions into the driver body, so we can treat some components of the binding in a per-parameter way and some in a per-operation way, in whatever way is better.

### Details and comments

Fix #10282
Fix #10283

As an example, take a setup of
```python
import math
from qiskit.circuit.library import EfficientSU2

qc = EfficientSU2(100, entanglement="linear", reps=100)
qc.measure_all()
ps = {x: math.pi / 2 for x in qc.parameters}
```
then the binding example
```python
qc.assign_parameters(ps)
```
(or `bind_parameters`) took my machine about 7m10s on `main`, and 1.24s after this PR (so a ~350x speedup).  It's probably not quite as fast as the flattening in #10269 because we still have _some_ overhead from dealing with the extra structure, but it's a big general step in a good direction, with (hopefully) no other user-facing implications.

### CODE REVIEW

File: test/python/qasm3/test_export.py
Comment (jakelishman) [inline test/python/qasm3/test_export.py:1]: The changed tests in this file are indicative of a problem in the parameter binding of prior internal custom operations.  The old tests were enforcing behaviour that shouldn't have been expected, given the current difficulties in looking at parametrised definitions after they've been bound - the `p0` and `p1` names shouldn't have been visible in the current data model (though in our preferred lazily bound definitions model, they would actually become visible).

File: qiskit/circuit/equivalence.py [hunk 1]
```diff
@@ -284,7 +285,7 @@ def _raise_if_shape_mismatch(gate, circuit):
 
 def _rebind_equiv(equiv, query_params):
     equiv_params, equiv_circuit = equiv
-    param_map = dict(zip(equiv_params, query_params))
-    equiv = equiv_circuit.assign_parameters(param_map, inplace=False)
+    param_map = dict((x, y) for x, y in zip(equiv_params, query_params) if isinstance(x, Parameter))
+    equiv = equiv_circuit.assign_parameters(param_map, inplace=False, flat_input=True)
```
Comment (jakelishman) [inline qiskit/circuit/equivalence.py:289]: This was a logical bug in the `EquivalenceLibrary` code that previously was being masked by a side-effect of `QuantumCircuit._unroll_params_dict` - it didn't raise errors on bad parameters in the map, it just silently dropped them.  For equivalences that have a mixture of symbolic and concrete parameters (there's at least one explicitly tested), this resulted in passing a map that looked like `{Parameter: float, float: float}`, which should be a typing error - a `float` isn't a valid `Parameter` and can't be bound.

**Review by Cryoris — COMMENTED** [review]
This is a massive speedup for nested circuits 👍🏻 But for already unrolled circuits, that need no rebinding, this change seems to be a bit slower than before. We could run some more benchmarks, but for PEC-style circuits I observed the following
```
Rebinding needed (circuit not manually decomposed)
----------------
main: 321.2s
this PR: 30.3s (10x faster)

Circuit unrolled
----------------
main: 10.4s +- 0.3
this PR: 12.9s +- 0.4 (~20% slower)
```

Do you have an idea where this slowdown might come from?

File: qiskit/circuit/quantumcircuit.py [hunk 2]
```diff
@@ -2737,47 +2753,148 @@ def assign_parameters(
...
+                        if new_parameter.is_real():
+                            new_parameter = (
+                                int(new_parameter)
+                                if new_parameter._symbol_expr.is_integer
+                                else float(new_parameter)
+                            )
+                        else:
+                            new_parameter = complex(new_parameter)
+                        new_parameter = operation.validate_parameter(new_parameter)
+                elif isinstance(assignee, QuantumCircuit):
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py:2827]: Control-flow ops (though I wasn't a huge fan of that).

**Comment by jakelishman** [comment]
For the PEC slowdown: thanks for reminding me.  I had seen a small slowdown, but not of the same magnitude as you. The reason is likely because the PEC code passes a vector of parameters to bind rather than a dictionary, and there's a bit more overhead in the new code to normalise the input formats.  I should be able to fix that.

**Comment by jakelishman** [comment]
I've done a little bit of lazy-evaluation trickery to remove the overhead from input normalisation, so nothing gets calculated until it's actually used.  Stemming from that abstraction, I was able to remove some overhead from the dictionary input form as well (no need to make a separate filtered dictionary; the filtering is just done as part of the iteration).

**Review by Cryoris — APPROVED** [review]
LGTM, it's great that this also provides a speedup in the unrolled case now 🙂

**Pull request merged**